README.md
<a href="https://www.gotoiot.com/">
    <img src="doc/gotoiot-logo.png" alt="logo" title="Goto IoT" align="right" width="60" height="60" />
</a>

Web App Full Stack Base
=======================

*Ayudaría mucho si apoyaras este proyecto con una ⭐ en Github!*

Este proyecto es una aplicación web fullstack que se ejecuta sobre el ecosistema `Docker`. Está compuesta por un compilador de `TypeScript` que te permite utilizar este superset de JavaScript para poder programar un `cliente web`. También tiene un servicio en `NodeJS` que te permite ejecutar código en backend y al mismo tiempo disponibilizar el código del cliente web para interactar con el servicio. Además tiene una `base de datos` MySQL que puede interactuar con el backend para guardar y consultar datos, y de manera adicional trae un `administrador` de base de datos para poder administrar la base en caso que lo necesites.

La aplicación IoT de base que viene con este proyecto se encarga de crear una tabla llamada `Devices` en la base de datos, y la idea es que vos puedas desarrollar el código de backend y frontend que te permita controlar desde el navegador el estado de los devices de un hogar inteligente - *como pueden ser luces, TVs, ventiladores, persianas, enchufes y otros* - y almacenar los estados de cada uno en la base de datos. 

Realizando estas tareas vas a a tener una aplicación fullstack IoT del mundo real que utiliza tecnologías actuales en la que un backend es capaz de interactuar con una DB para cumplir con las peticiones de control que se le mandan desde el cliente web.

En esta imagen podés ver una posible implementación del cliente web que controla los artefactos del hogar.

![architecture](doc/webapp-example-1.png)

## Comenzando 🚀

Esta sección es una guía con los pasos escenciales para que puedas poner en marcha la aplicación.

<details><summary><b>Mira los pasos necesarios</b></summary><br>

### Instalar las dependencias

Para correr este proyecto es necesario que instales `Docker` y `Docker Compose`. 

En [este artículo](https://www.gotoiot.com/pages/articles/docker_installation_linux/) publicado en nuestra web están los detalles para instalar Docker y Docker Compose en una máquina Linux. Si querés instalar ambas herramientas en una Raspberry Pi podés seguir [este artículo](https://www.gotoiot.com/pages/articles/rpi_docker_installation) de nuestra web que te muestra todos los pasos necesarios.

En caso que quieras instalar las herramientas en otra plataforma o tengas algún incoveniente, podes leer la documentación oficial de [Docker](https://docs.docker.com/get-docker/) y también la de [Docker Compose](https://docs.docker.com/compose/install/).

Continua con la descarga del código cuando tengas las dependencias instaladas y funcionando.

### Descargar el código

Para descargar el código, lo más conveniente es que realices un `fork` de este proyecto a tu cuenta personal haciendo click en [este link](https://github.com/gotoiot/app-fullstack-base/fork). Una vez que ya tengas el fork a tu cuenta, descargalo con este comando (acordate de poner tu usuario en el link):

```
git clone https://github.com/USER/app-fullstack-base.git
```

> En caso que no tengas una cuenta en Github podes clonar directamente este repo.

### Ejecutar la aplicación

Para ejecutar la aplicación tenes que correr el comando `docker-compose up` desde la raíz del proyecto. Este comando va a descargar las imágenes de Docker de node, de typescript, de la base datos y del admin de la DB, y luego ponerlas en funcionamiento. 

Para acceder al cliente web ingresa a a la URL [http://localhost:8000/](http://localhost:8000/) y para acceder al admin de la DB accedé a [localhost:8001/](http://localhost:8001/). 

Si pudiste acceder al cliente web y al administrador significa que la aplicación se encuentra corriendo bien. 

> Si te aparece un error la primera vez que corres la app, deteńe el proceso y volvé a iniciarla. Esto es debido a que el backend espera que la DB esté creada al iniciar, y en la primera ejecución puede no alcanzar a crearse. A partir de la segunda vez el problema queda solucionado.

</details>

Continuá explorando el proyecto una vez que lo tengas funcionando.

## Configuraciones de funcionamiento 🔩

Al crearse la aplicación se ejecutan los contenedores de Docker de cada servicio, se crea la base de datos y sus tablas. A continuación podés encontrar info si querés cambiar la estructura de la DB o bien sus configuraciones de acceso.

<details><summary><b>Lee cómo configurar la aplicación</b></summary><br>

### Configuración de la DB

Como ya comprobaste, para acceder PHPMyAdmin tenés que ingresar en la URL [localhost:8001/](http://localhost:8001/). En el login del administrador, el usuario para acceder a la db es `root` y contraseña es la variable `MYSQL_ROOT_PASSWORD` del archivo `docker-compose.yml`.

Para el caso del servicio de NodeJS que se comunica con la DB fijate que en el archivo `src/backend/mysql-connector.js` están los datos de acceso para ingresar a la base.

Si quisieras cambiar la contraseña, puertos, hostname u otras configuraciones de la DB deberías primero modificar el servicio de la DB en el archivo `docker-compose.yml` y luego actualizar las configuraciones para acceder desde PHPMyAdmin y el servicio de NodeJS.

### Estructura de la DB

Al iniciar el servicio de la base de datos, si esta no está creada toma el archivo que se encuentra en `db/dumps/smart_home.sql` para crear la base de datos automáticamente.

En ese archivo está la configuración de la tabla `Devices` y otras configuraciones más. Si quisieras cambiar algunas configuraciones deberías modificar este archivo y crear nuevamente la base de datos para que se tomen en cuenta los cambios.

Tené en cuenta que la base de datos se crea con permisos de superusuario por lo que no podrías borrar el directorio con tu usuario de sistema, para eso debés hacerlo con permisos de administrador. En ese caso podés ejecutar el comando `sudo rm -r db/data` para borrar el directorio completo.

</details>


## Detalles principales 🔍

En esta sección vas a encontrar las características más relevantes del proyecto.

<details><summary><b>Mira los detalles más importantes de la aplicación</b></summary><br>
<br>

### Arquitectura de la aplicación

Como ya pudiste ver, la aplicación se ejecuta sobre el ecosistema Docker, y en esta imagen podés ver el diagrama de arquitectura.

![architecture](doc/architecture.png)

### El cliente web

El cliente web es una Single Page Application que se comunica con el servicio en NodeJS mediante JSON a través de requests HTTP. Puede consultar el estado de dispositivos en la base de datos (por medio del servicio en NodeJS) y también cambiar el estado de los mismos. Los estilos del código están basados en **Material Design**.

### El servicio web

El servicio en **NodeJS** posee distintos endpoints para comunicarse con el cliente web mediante requests HTTP enviando **JSON** en cada transacción. Procesando estos requests es capaz de comunicarse con la base de datos para consultar y controlar el estado de los dispositivos, y devolverle una respuesta al cliente web también en formato JSON. Así mismo el servicio es capaz de servir el código del cliente web.

### La base de datos

La base de datos se comunica con el servicio de NodeJS y permite almacenar el estado de los dispositivos en la tabla **Devices**. Ejecuta un motor **MySQL versión 5.7** y permite que la comunicación con sus clientes pueda realizarse usando usuario y contraseña en texto plano. En versiones posteriores es necesario brindar claves de acceso, por este motivo la versión 5.7 es bastante utilizada para fases de desarrollo.

### El administrador de la DB

Para esta aplicación se usa **PHPMyAdmin**, que es un administrador de base de datos web muy utilizado y que podés utilizar en caso que quieras realizar operaciones con la base, como crear tablas, modificar columnas, hacer consultas y otras cosas más.

### El compilador de TypeScript

**TypeScript** es un lenguaje de programación libre y de código abierto desarrollado y mantenido por Microsoft. Es un superconjunto de JavaScript, que esencialmente añade tipos estáticos y objetos basados en clases. Para esta aplicación se usa un compilador de TypeScript basado en una imagen de [Harmish](https://hub.docker.com/r/harmish) en Dockerhub, y está configurado para monitorear en tiempo real los cambios que se realizan sobre el directorio **src/frontend/ts** y automáticamente generar código compilado a JavaScript en el directorio  **src/frontend/js**. Los mensajes del compilador aparecen automáticamente en la terminal al ejecutar el comando **docker-compose up**.

### Ejecución de servicios

Los servicios de la aplicación se ejecutan sobre **contenedores de Docker**, así se pueden desplegar de igual manera en diferentes plataformas. Los detalles sobre cómo funcionan los servicios los podés ver directamente en el archivo **docker-compose.yml**.

### Organización del proyecto

En la siguiente ilustración podés ver cómo está organizado el proyecto para que tengas en claro qué cosas hay en cada lugar.

```sh
├── db                          # directorio de la DB
│   ├── data                    # estructura y datos de la DB
│   └── dumps                   # directorio de estructuras de la DB
│       └── smart_home.sql      # estructura con la base de datos "smart_home"
├── doc                         # documentacion general del proyecto
└── src                         # directorio codigo fuente
│   ├── backend                 # directorio para el backend de la aplicacion
│   │   ├── index.js            # codigo principal del backend
│   │   ├── mysql-connector.js  # codigo de conexion a la base de datos
│   │   ├── package.json        # configuracion de proyecto NodeJS
│   │   └── package-lock.json   # configuracion de proyecto NodeJS
│   └── frontend                # directorio para el frontend de la aplicacion
│       ├── js                  # codigo javascript que se compila automáticamente
│       ├── static              # donde alojan archivos de estilos, imagenes, fuentes, etc.
│       ├── ts                  # donde se encuentra el codigo TypeScript a desarrollar
│       └── index.html          # archivo principal del cliente HTML
├── docker-compose.yml          # archivo donde se aloja la configuracion completa
├── README.md                   # este archivo
├── CHANGELOG.md                # archivo para guardar los cambios del proyecto
├── LICENSE.md                  # licencia del proyecto
```

> No olvides ir poniendo tus cambios en el archivo `CHANGELOG.md` a medida que avanzas en el proyecto.

</details>

## Detalles de implementación 💻

En esta sección podés ver los detalles específicos de funcionamiento del código y que son los siguientes.

<details><summary><b>Mira los detalles de implementación</b></summary><br>

### Obtener dispositivos

Al inicializarse la pagina se listan todos los dispositivos del sistema, por cada uno se dibuja una card en la cual se muestra el nombre, el detalle el cual puede verse clickeando en los 3 puntos y las opciones de actualizar, eliminar y cambiar de estado.

### Agregar un dispositivo

Para agregar un dispositivo nuevo hacer click en el boton "Agregar" qu se encuentra en la parte superior. Se abrira una 
ventana modal en la que se ingresará:
    -Nombre
    -Descripcion
    -Tipo (Regulable, ON/OFF)
    -Por defecto el estado se crea en 0

### Actualizar un dispositivo

Para actualizar los datos en cada card se encuentra la opcion "Actualizar", al presionarla se abrira una modal que nos permite modificar: Nombre, Descripcion y tipo, luego confgirmar

### Actualizar estado de un dispositivo
En cada card, para cada device, se encuentra un switch (ON/OFF o un slider) segun el tipo de dispositivo. Luego de cambiar el valor se debe confirmar con el boton que tiene el icono de un tilde

### Eliminar un dispositivo
En cada card, para cada dispositivo, se encuentra el boton eliminar, esto eliminara del sistema al device


### Frontend

El frontend está desarrollado usando HTML, typescript y css. Se utilizan elementos de materialize (cards, modals y floatting buttons). La conunicación con el
backend es mediante AJAX para las llamadas a las api REST y se usan listeners para procesar las respuestas

### Backend

La base de datos mysql permite almacenar los diferentes devices del sistema y se permite obtener funcionalidad mediante una api rest realizada con express y node js.   Para la interaccion con la base se utiliza la clase mysql-connector que nos brinda la conexion hacia la base y nos permite ejecutar sentencias SQL

<details><summary><b>Ver los endpoints disponibles</b></summary><br>

Completá todos los endpoints del backend con los metodos disponibles, los headers y body que recibe, lo que devuelve, ejemplos, etc.

1) Devolver el estado de los dispositivos.

URL: /devices
Metodo: GET

 Recibe opcionalmente parametro para ordenar ascendente/descendente
 
 Codigos de estado:
  -500: Error interno al conectarse a la base
  -200: Funcionamiento correcto
 
  Body Response:
   Array de json con los datos de cada device:
       [{ id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)}]

2) Devolver el estado de un dispositivo.

URL: /devices:id
Metodo: GET

  PathParam: Id del dispositivo
  
  Codigos de estado:
   -500: Error interno al conectarse a la base
   -200: Funcionamiento correcto
   -404: No existe el device
  
  Body Response:
   Json con los datos de cada device:
        {id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)}

3) Eliminar un device.

URL: /devices:id
Metodo: DELETE

 PathParam: Id del dispositivo
 
  Codigos de estado:
   -500: Error interno al conectarse a la base
   -200: Funcionamiento correcto
   -404: No existe el device
  
  Body Response:
   empty

4) Crear un device.

URL: /devices
Metodo: POST

  
  Codigos de estado:
   -500: Error interno al conectarse a la base
   -201: Se creo un device exitosamente
  Request Body:
        {name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)}
  
  Body Response:
   Json con los datos del device creado:
        {id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)}
5) Actualizar datos de un device.

URL: /devices/:id
Metodo: PUT

 PathParam: Id del dispositivo    
   Codigos de estado:
   -500: Error interno al conectarse a la base
   -404: No existe device
   -200: Se actualizo un device exitosamente
  Request Body:
        {name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)}
  
  Response Body:
   Json con los datos del device actualizado:
        {id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)}
6) Actualizar el estado de un device                           
  URL: device/:id/state
  Metodo: PATCH
  
  PathParam: Id del dispositivo
  Codigos de estado:
   -500: Error interno al conectarse a la base
   -404: No existe device
   -200: Se actualizo un device exitosamente
  Request Body:
        state: string
        
  
  Response Body:
   Json con los datos del device actualizado:
        id: number, 
         state: number

</details>

</details>


## Tecnologías utilizadas 🛠️

En esta sección podés ver las tecnologías más importantes utilizadas.

<details><summary><b>Mira la lista completa de tecnologías</b></summary><br>

* [Docker](https://www.docker.com/) - Ecosistema que permite la ejecución de contenedores de software.
* [Docker Compose](https://docs.docker.com/compose/) - Herramienta que permite administrar múltiples contenedores de Docker.
* [Node JS](https://nodejs.org/es/) - Motor de ejecución de código JavaScript en backend.
* [MySQL](https://www.mysql.com/) - Base de datos para consultar y almacenar datos.
* [PHPMyAdmin](https://www.phpmyadmin.net/) - Administrador web de base de datos.
* [Material Design](https://material.io/design) - Bibliotecas de estilo responsive para aplicaciones web.
* [TypeScript](https://www.typescriptlang.org/) - Superset de JavaScript tipado y con clases.

</details>

## Contribuir 🖇️

Si estás interesado en el proyecto y te gustaría sumar fuerzas para que siga creciendo y mejorando, podés abrir un hilo de discusión para charlar tus propuestas en [este link](https://github.com/gotoiot/app-fullstack-base/issues/new). Así mismo podés leer el archivo [Contribuir.md](https://github.com/gotoiot/gotoiot-doc/wiki/Contribuir) de nuestra Wiki donde están bien explicados los pasos para que puedas enviarnos pull requests.

## Sobre Goto IoT 📖

Goto IoT es una plataforma que publica material y proyectos de código abierto bien documentados junto a una comunidad libre que colabora y promueve el conocimiento sobre IoT entre sus miembros. Acá podés ver los links más importantes:

* **[Sitio web](https://www.gotoiot.com/):** Donde se publican los artículos y proyectos sobre IoT. 
* **[Github de Goto IoT:](https://github.com/gotoiot)** Donde están alojados los proyectos para descargar y utilizar. 
* **[Comunidad de Goto IoT:](https://groups.google.com/g/gotoiot)** Donde los miembros de la comunidad intercambian información e ideas, realizan consultas, solucionan problemas y comparten novedades.
* **[Twitter de Goto IoT:](https://twitter.com/gotoiot)** Donde se publican las novedades del sitio y temas relacionados con IoT.
* **[Wiki de Goto IoT:](https://github.com/gotoiot/doc/wiki)** Donde hay información de desarrollo complementaria para ampliar el contexto.

## Muestas de agradecimiento 🎁

Si te gustó este proyecto y quisieras apoyarlo, cualquiera de estas acciones estaría más que bien para nosotros:

* Apoyar este proyecto con una ⭐ en Github para llegar a más personas.
* Sumarte a [nuestra comunidad](https://groups.google.com/g/gotoiot) abierta y dejar un feedback sobre qué te pareció el proyecto.
* [Seguirnos en twitter](https://github.com/gotoiot/doc/wiki) y dejar algún comentario o like.
* Compartir este proyecto con otras personas.

## Autores 👥

Las colaboraciones principales fueron realizadas por:

* **[Agustin Bassi](https://github.com/agustinBassi)**: Ideación, puesta en marcha y mantenimiento del proyecto.
* **[Ernesto Giggliotti](https://github.com/ernesto-g)**: Creación inicial del frontend, elección de Material Design.
* **[Brian Ducca](https://github.com/brianducca)**: Ayuda para conectar el backend a la base de datos, puesta a punto de imagen de Docker.

También podés mirar todas las personas que han participado en la [lista completa de contribuyentes](https://github.com/###/contributors).

## Licencia 📄

Este proyecto está bajo Licencia ([MIT](https://choosealicense.com/licenses/mit/)). Podés ver el archivo [LICENSE.md](LICENSE.md) para más detalles sobre el uso de este material.

---

**Copyright © Goto IoT 2021** ⌨️ [**Website**](https://www.gotoiot.com) ⌨️ [**Group**](https://groups.google.com/g/gotoiot) ⌨️ [**Github**](https://www.github.com/gotoiot) ⌨️ [**Twitter**](https://www.twitter.com/gotoiot) ⌨️ [**Wiki**](https://github.com/gotoiot/doc/wiki)


CHANGELOG.md
# DAW Base App - Changes Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 2.3.0

* Project modification
    * Adds endpoints to create, update, delete and get devices
    * Adds sql querys to get data
  


## 2.2.0

* Project modification
    * Adds TypeScript compiler service to Docker Compose
    * Reestructures frontend folder for TypeScript
    * Adds new info to README accordingly
    * Changes project architecture image

## 2.1.0

* Project modification
    * Enhaces README accordingly to Goto IoT
    * Adds example of finished application
    * Removes unnecessary frontend images
    * Changes src code folders names

## 2.0.0

* Project modification
    * Changes project and organization names
    * Removes Typescript container
    * Removes Typescript Code
    * Executes Javascript code directly
    * Changes licence to MIT
    * Modifies README accordingly

## 1.0.0

* Project creation
    * Docker Compose implementation for whole project.
    * Typescript compilation into docker-compose.
    * MySQL 5.7 DB Server.
    * PHPMyAdmin.
    * NodeJS backend application.
    * Materialize CSS framework.


src/frontend/ts/RestFramework.ts
/**
 *  Clase que encapsula las llamadas Rest de AJAX
 *  Se agregan interfaces para los listener para las respuestas de las llamadas
 * 
 * 
 */


class RestFramework {


    public requestGET(url:string, listener:GETResponseListener){
        let xml = new XMLHttpRequest();

        xml.onreadystatechange = function respuestaServidor(){
            
            if(xml.readyState == 4)
            {
                listener.handleGETResponse(xml.status, xml.response);  
            }                        

        }
        xml.open("GET", url,true);//true --->asincrona
        xml.send();
   }

   public requestDEL(url:string, listener:DELETEResponseListener){
       ///AJAX api rest DEL
       let xml = new XMLHttpRequest();

       //asynchronous request method
       xml.onreadystatechange = function respuestaServidor(){
           
           if(xml.readyState == 4)//status 4 all transaction performed
           {
               listener.handleDELETEResponse(xml.status, xml.response);      
           }                        
       }
       xml.open("DELETE", url,true);//true --->asincrona
       xml.send();
  }

  public requestPOST(url:string, listener:POSTResponseListener,data:any){
       let xml = new XMLHttpRequest();
       //asynchronous request method
       let jdata=JSON.stringify(data);
       xml.onreadystatechange = function respuestaServidor(){
           
           if(xml.readyState == 4)//status 4 all transaction performed
           {
               listener.handlePOSTResponse(xml.status, xml.response);      
           }                        
       }
       xml.open("POST", url,true);//true --->asincrona
       xml.setRequestHeader("Content-Type","application/json;charset=UTF-8");
       xml.send(jdata);
   }


   public requestPATCH(url:string, listener:PATCHResponseListener, data:any){
    let xml = new XMLHttpRequest();
    //asynchronous request method
    let jdata=JSON.stringify(data);   
    xml.onreadystatechange = function respuestaServidor(){
   
        if(xml.readyState == 4)//status 4 all transaction performed
        {
            listener.handlePATCHResponse(xml.status, xml.response);      
        }                        
    }
    xml.open("PATCH", url,true);//true --->asincrona
    xml.setRequestHeader("Content-Type","application/json;charset=UTF-8");
    xml.send(jdata);
}




   public requestPUT(url:string, listener:PutResponseListener,data:any){
   ///AJAX api rest PUT for state changes
   console.log("sending state changes")
   let xml = new XMLHttpRequest();
   console.log(data);
   //asynchronous request method
   let jdata=JSON.stringify(data);
   console.log(jdata);
   xml.onreadystatechange = function respuestaServidor(){
       
       if(xml.readyState == 4)//status 4 all transaction performed
       {
           listener.handlePutResponse(xml.status, xml.response);      
       }                        
   }
   xml.open("PUT", url,true);//true --->asincrona
   xml.setRequestHeader("Content-Type","application/json;charset=UTF-8");
   xml.send(jdata);

   }
}

interface GETResponseListener {
    handleGETResponse (status: number, response: string): void;
}

interface POSTResponseListener {
    handlePOSTResponse (status: number, response: string): void;
}

interface DELETEResponseListener {
    handleDELETEResponse (status: number, response: string): void;
}

interface PATCHResponseListener {
    handlePATCHResponse (status: number, response: string): void;
}

interface PutResponseListener {
    handlePutResponse (status: number, response: string): void;
}

src/backend/index.js
//=======[ Settings, Imports & Data ]==========================================

var PORT    = 3000;

var express = require('express');
var app     = express();
var utils   = require('./mysql-connector');

// to parse application/json
app.use(express.json()); 
// to serve static files
app.use(express.static('/home/node/app/static/'));

//=======[ Main module code ]==================================================

jsonDevices = require('./devices.js')

/**
 * Metodo que captura request para obtener todos los devices
 * Recibe opcionalmente parametro para ordenar ascendente/descendente
 * 
 * Codigos de estado:
 *  -500: Error interno al conectarse a la base
 *  -200: Funcionamiento correcto
 * 
 * Body:
 *  Array de json con los datos de cada device:
 *       id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)
 * 
 */
app.get('/devices/', function(req, res, next) {
    let query = "SELECT * FROM Devices";
    if( req.query.sort === 'DES') {
       query += " ORDER BY id DESC";
    } else {
        query += " ORDER BY id ASC";     
    }
    utils.query(query, (error,result, fields) =>{
        if(error) return res.status(500).send("Error:"+ error);
        return res.status(200).send(JSON.stringify(result));
    }); 
});


/**
 * Metodo que captura request para obtener un device
 * PathParam: Id del dispositivo
 * 
 * Codigos de estado:
 *  -500: Error interno al conectarse a la base
 *  -200: Funcionamiento correcto
 *  -404: No existe el device
 * 
 * Body:
 *  Json con los datos de cada device:
 *       id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)
 * 
 */

app.get('/devices/:id', function(req, res, next) {
/*jsonDevices.forEach(element => {
        if(element.id == req.params.id ) {
            return res.status(200).send(JSON.stringify(element));
        } 
    });
    
    return res.status(404).send(null);
*/
    let query = "SELECT * FROM Devices WHERE id = ?";
    utils.query(query, [req.params.id], (error,result, fields) => {
        if(error) return res.status(500).send("Error:"+ error);
        if(result.affectedRows == 0) return res.status(404).send(null);
        return res.status(200).send(JSON.stringify(result));
    });
});

/**
 * Metodo que captura request para eliminar un device
 * PathParam: Id del dispositivo
 * 
 * Codigos de estado:
 *  -500: Error interno al conectarse a la base
 *  -200: Funcionamiento correcto
 *  -404: No existe el device
 * 
 * Body:
 *  empty
 * 
 */
app.delete('/devices/:id', function(req, res, next) {
   /* jsonDevices.forEach( (element, index) => {
           if(element.id == req.params.id ) {
                jsonDevices.splice(index,1)
                return res.status(200).send(JSON.stringify(element));
           } 
       });*/
        
     //  return res.status(404).send(null);
     let deleteSQL = "DELETE FROM Devices WHERE id = ?" ;
    
     utils.query(deleteSQL, [req.params.id],(error, result, fields) => {
        if(error) return res.status(500).send("Error:"+ error);
        if(result.affectedRows == 0) return res.status(404).send(null); 
        return res.status(200).send();
     });
   });


   /**
 * Metodo que captura request para crear device
 * 
 * Codigos de estado:
 *  -500: Error interno al conectarse a la base
 *  -201: Se creo un device exitosamente
 * Request Body:
 *       name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)
 * 
 * Body:
 *  Json con los datos del device creado:
 *       id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)
 * 
 */
app.post('/devices', function (req, res, next) {
    
  //  const DeviceDTO = require('./devicedto.js')
  /*  const device = new DeviceDTO(req.body.id, req.body.name, req.body.description, req.body.state, req.body.type);
    if( jsonDevices.some( element => element.id === device.id )) {
        return res.status(400).send("Ya existe un device con ese id");
    }
    jsonDevices.push(device);*/
    let insertSQL = "INSERT INTO Devices (name,description,state,type) VALUES (?,?,?,?)";
    
    utils.query(insertSQL, [req.body.name, req.body.description, req.body.state, req.body.type],(error, result, fields) => {
        if(error) return res.status(500).send("Error:"+ error);
        return res.status(201).send(req.body);
    });

    


});

/**
 * Metodo que captura request para actualizar device
 * 
 * Codigos de estado:
 *  -500: Error interno al conectarse a la base
 *  -404: No existe device
 *  -200: Se actualizo un device exitosamente
 * Request Body:
 *       name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)
 * 
 * Body:
 *  Json con los datos del device actualizado:
 *       id: number, 
        name: string, 
        description: string, 
        state: number, 
        type: number (1-On/Off, 2-Regulable)
 * 
 */

app.put('/devices/:id', function (req, res, next) {
    
    /*
    jsonDevices.forEach( (element, index) => {
        if(element.id == req.params.id ) {
             element.name = req.body.name;
             element.description = req.body.description;
             element.state = req.body.state;
             element.type = req.body.type;
             jsonDevices.splice(index,1);
             jsonDevices.push(element);
             return res.status(200).send(JSON.stringify(element));
        } 
    });*/
    let updateSQL = "UPDATE Devices SET name =?,description =?,state =?,type=? WHERE ID = ?";
    
    utils.query(updateSQL, [req.body.name, req.body.description, req.body.state, req.body.type,req.params.id],(error, result, fields) => {
        if(error) return res.status(500).send("Error:"+ error);
        if(result.affectedRows == 0) return res.status(404).send(null);
        
        return res.status(200).send(JSON.stringify(req.body));
    });
    


});

/**
 * Metodo que captura request para actualizar el estado de un device
 * 
 * Codigos de estado:
 *  -500: Error interno al conectarse a la base
 *  -404: No existe device
 *  -200: Se actualizo un device exitosamente
 * Request Body:
 *       state: string
        
 * 
 * Body:
 *  Json con los datos del device actualizado:
 *       id: number, 
         state: number
 * 
 */

app.patch('/devices/:id/state', function (req, res, next) {
    
    /*
    jsonDevices.forEach( (element, index) => {
        if(element.id == req.params.id ) {
             
             element.state = req.body.state;
             jsonDevices.splice(index,1);
             jsonDevices.push(element);
             return res.status(200).send(JSON.stringify(element));
        } 
    });
    
    return res.status(404).send(null);
    */
   let updateSQL = "UPDATE Devices SET state =? WHERE ID = ?";
   utils.query(updateSQL, [req.body.state, req.params.id],(error, result, fields) => {
    if(error) return res.status(500).send("Error:"+ error);
    if(result.affectedRows == 0) return res.status(400).send(null);   
    
    return res.status(200).send(JSON.stringify({id:req.params.id, state: req.body.state}));
   });

});

app.listen(PORT, function(req, res) {
    console.log("NodeJS API running correctly");
});


//=======[ End of file ]=======================================================
